lib/elixir_boilerplate_web/errors/helpers.ex

```diff
@@ -1,43 +1,47 @@
 defmodule ElixirBoilerplateWeb.Errors.Helpers do
-  @moduledoc """
-  Conveniences for translating and building error messages.
-  """
-  use Phoenix.HTML
+  alias Ecto.Changeset
+  alias ElixirBoilerplateWeb.Errors.View
 
   @doc """
-  Generates tag for inlined form input errors.
+  Generates a human-readable block containing all errors in a changeset
   """
-  def error_tag(form, field) do
-    Enum.map(Keyword.get_values(form.errors, field), fn error ->
-      content_tag(:span, translate_error(error))
-    end)
+  def error_messages(changeset) do
+    changeset
+    |> Changeset.traverse_errors(&translate_error/1)
+    |> convert_errors_to_html(changeset.data.__struct__)
   end
 
-  @doc """
-  Translates an error message using gettext.
-  """
-  def translate_error({msg, opts}) do
-    # When using gettext, we typically pass the strings we want
-    # to translate as a static argument:
-    #
-    #     # Translate "is invalid" in the "errors" domain
-    #     dgettext("errors", "is invalid")
-    #
-    #     # Translate the number of files with plural rules
-    #     dngettext("errors", "1 file", "%{count} files", count)
-    #
-    # Because the error messages we show in our forms and APIs
-    # are defined inside Ecto, we need to translate them dynamically.
-    # This requires us to call the Gettext module passing our gettext
-    # backend as first argument.
-    #
-    # Note we use the "errors" domain, which means translations
-    # should be written to the errors.po file. The :count option is
-    # set by Ecto and indicates we should also apply plural rules.
-    if count = opts[:count] do
-      Gettext.dngettext(ElixirBoilerplateWeb.Gettext, "errors", msg, msg, count, opts)
+  defp translate_error({message, options}) do
+    if options[:count] do
+      Gettext.dngettext(ElixirBoilerplate.Gettext, "ecto", message, message, options[:count], options)
     else
-      Gettext.dgettext(ElixirBoilerplateWeb.Gettext, "errors", msg, opts)
+      Gettext.dgettext(ElixirBoilerplate.Gettext, "ecto", message, options)
     end
   end
+
+  defp convert_errors_to_html(errors, schema) do
+    errors = Enum.reduce(errors, [], &convert_error_field(&1, &2, schema))
+    View.render("error_messages.html", %{errors: errors})
+  end
+
+  defp convert_error_field({field, errors}, memo, schema) when is_list(errors), do: memo ++ Enum.flat_map(errors, &convert_error_subfield(&1, field, [], schema))
+  defp convert_error_field({field, errors}, memo, schema) when is_map(errors), do: memo ++ Enum.flat_map(Map.keys(errors), &convert_error_subfield(&1, field, errors[&1], schema))
+
+  defp convert_error_subfield(message, field, _, _schema) when is_binary(message) do
+    # NOTE `schema` is available here if we want to use something like
+    # `schema.humanize_field(field)` to be able to display `"Email address is
+    # invalid"` instead of `email is invalid"`.
+    ["#{field} #{message}"]
```

Comment:
I like the idea here. It would be nice to cover this as an example through tests.

Reply:
You mean the example using a `schema.humanize_field/1` function?

It would be kind of hard to _test_ it because we would have to actually use it here in the helper 😕